components/script/dom/transformstreamdefaultcontroller.rs

```diff
@@ -51,22 +51,130 @@ impl Callback for TransformTransformPromiseRejection {
     }
 }
 
+pub(crate) trait TransformerCancelAlgorithm: Trace {
+    fn run(
+        &self,
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        chunk: SafeHandleValue,
+        can_gc: CanGc,
+    ) -> Fallible<Rc<Promise>>;
+}
+
+#[allow(unused)]
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+pub(crate) enum TransformerCancelAlgorithmType {
+    Js(#[ignore_malloc_size_of = "Rc is hard"] Rc<TransformerCancelCallback>),
+    Native(#[ignore_malloc_size_of = "Rc is hard"] Rc<dyn TransformerCancelAlgorithm>),
+}
+
+impl TransformerCancelAlgorithmType {
+    fn call(
+        &self,
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        this_obj: &SafeHandleObject,
+        chunk: SafeHandleValue,
+        can_gc: CanGc,
+    ) -> Fallible<Rc<Promise>> {
+        match self {
+            TransformerCancelAlgorithmType::Js(cancel) => {
+                cancel.Call_(this_obj, chunk, ExceptionHandling::Rethrow, can_gc)
+            },
+            TransformerCancelAlgorithmType::Native(cancel) => cancel.run(cx, global, chunk, can_gc),
+        }
+    }
+}
+
+pub(crate) trait TransformerFlushAlgorithm: Trace {
+    fn run(
+        &self,
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        controller: &TransformStreamDefaultController,
+        can_gc: CanGc,
+    ) -> Fallible<Rc<Promise>>;
+}
+
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+pub(crate) enum TransformerFlushAlgorithmType {
+    Js(#[ignore_malloc_size_of = "Rc is hard"] Rc<TransformerFlushCallback>),
+    Native(#[ignore_malloc_size_of = "Rc is hard"] Rc<dyn TransformerFlushAlgorithm>),
```

Comment:
So you don't need one type per algorithm, you just need on type per transformer. 

Example:

```
enum TransformerType {
      Js {
           flush: Rc<TransformerFlushCallback>,
           // the others as well.
      },
     Decoder(Rc<TextDecoderCommon>)
}
```

This is not a store of algorithms, it's a store of data that are necessary to run the algorithms in the controller itself. So for example, in `perform_transform`, you would use the decoder stored in the enum to run the algorithm as Rust code defined in the match block.